Differentiate between translating libraries and executables

alevy opened the pull request:
This PR adds a `Tool` and `Representation` that identify based on the `CMakeLists.txt` in the source project whether the target is a library or executable. It also uses that new representation in the LLM translation to use different system prompts for each.

This proved to be useful to help encourage the LLM to generate libraries that actually expose non-mangled C symbols and compile to `.so`s.

Finally, the PR also includes a `Cargo.lock` and a new `default.nix` for use in building the container for submission because I needed that and didn't pull it out to a separate branch.

jrvanwhy commented on ir/src/fs.rs:
```diff
@@ -132,6 +132,28 @@ impl RawDir {
+            } else {
+                None
+            }
+        })
```
Nit: I prefer:

```rust
        match cur_dir.0.get(file_name) {
            Some(RawEntry::File(v)) => Some(v),
            _ => None,
        }
```

jrvanwhy replied:
Nit: I prefer

```suggestion
        match cur_dir.0.get(file_name) {
            Some(RawEntry::File(v)) => Some(v),
            _ => None,
        }
```

jrvanwhy commented on translate/src/tools/raw_source_to_cargo_llm/mod.rs:
```diff
@@ -27,7 +30,12 @@ impl Tool for RawSourceToCargoLlm {
+            .get_by_representation::<ProjectKind>()
+            .zip(context.ir.get_by_representation::<RawSource>())
+            .next()
+        {
```
Nit: I prefer:

```suggestion
        match (
            context.ir.get_by_representation::<ProjectKind>().next(),
            context.ir.get_by_representation::<RawSource>().next(),
        ) {
```

jrvanwhy commented on translate/src/tools/raw_source_to_cargo_llm/mod.rs:
```diff
@@ -58,8 +78,16 @@ impl Tool for RawSourceToCargoLlm {
+                ProjectKind::Library => {
+                    llm_builder = llm_builder.system(SYSTEM_PROMPT_LIBRARY);
+                }
+            }
```
Nit: I prefer

```suggestion
            llm_builder = match project_kind {
                ProjectKind::Executable => llm_builder.system(SYSTEM_PROMPT_EXECUTABLE),
                ProjectKind::Library => llm_builder.system(SYSTEM_PROMPT_LIBRARY),
            };
```

jrvanwhy commented on translate/src/tools/raw_source_to_cargo_llm/system_prompt_executable.txt:
```diff
@@ -1,4 +1,4 @@
-You are a code translation tool. You translate provided C projects into a Rust projects including Cargo manifest. You replicate comments on functions, methods, structs, and modules, and, if they are missing in the original C project, you write appropriate comments for the resulting Rust. For example, given the following prompt:
+You are a code translation tool. You translate provided C projects into a Rust projects including Cargo manifest. You functions, methods, structs, and modules but not comments. Do not include or write any new comments. You preserve external interfaces but internally use cannonical and safe Rust as much as possible. When asked to generate a structured JSON response, you always respond with valid J ...
```
"You functions"? Same in system_prompt_library.txt.

I guess it's an LLM so whatever works *shrug*.

jrvanwhy commented on translate/src/tools/identify_project_kind.rs:
```diff
@@ -0,0 +1,72 @@
+            ProjectKind::Library => std::fs::write(path.join("library"), []),
+            ProjectKind::Executable => std::fs::write(path.join("executable"), []),
+        }
+    }
```
Why use the file name to communicate the type rather than file contents? It seems to me you could skip implementing `materialize` entirely and rely on the default impl (which dumps the `Display` output into a file).

alevy replied:
Oh... yes, will do. Originally I was getting both the project kind and the name, and storing the name in the file. That turns out to be more tricky to get reliably and also not immediately used anyway, so I got rid of it for now. Eventually, there are also projects that output both a library and an executable and might have different names for each.

But for now, just the default implementation is best indeed.

enjhnsn2 reviewed (approved):
Eventually we probably want a mechanism for distinguishing between the two that doesn't rely on cmake (maybe just grep all the source for a main?), but thats not an immediate issue.

Overall, lgtm.

alevy commented:
@jrvanwhy I think I addressed all your comments. For `get_file` I just used an analogous interface to `set_file` with the same semantics for directory traversal now.
